server/hotrod/src/main/scala/org/infinispan/server/hotrod/iteration/IterationFilters.scala

```diff
@@ -0,0 +1,127 @@
+package org.infinispan.server.hotrod.iteration
+
+import java.io.{ObjectInput, ObjectOutput}
+
+import org.infinispan.commons.io.UnsignedNumeric
+import org.infinispan.commons.marshall.{AbstractExternalizer, Marshaller}
+import org.infinispan.distribution.DistributionManager
+import org.infinispan.distribution.ch.ConsistentHash
+import org.infinispan.factories.annotations.Inject
+import org.infinispan.filter.{AbstractKeyValueFilterConverter, KeyValueFilterConverter}
+import org.infinispan.metadata.Metadata
+import org.infinispan.server.hotrod._
+
+import scala.collection.JavaConversions._
+
+/**
+ * @author gustavonalle
+ * @since 8.0
+ */
+
+trait InjectableFilter[K, V, C] extends AbstractKeyValueFilterConverter[K, V, C] {
+   var cache: Cache = _
+
+   @Inject def injectDependencies(cache: Cache) {
+      this.cache = cache
+   }
+}
+
+class WrappedFilter[K, V, C, D](val outer: InjectableFilter[K, C, D], val inner: InjectableFilter[K, V, C]) extends InjectableFilter[K, V, D] {
+   @Inject
+   override def injectDependencies(cache: Cache): Unit = {
+      outer.injectDependencies(cache)
+      inner.injectDependencies(cache)
+   }
+
+   override def filterAndConvert(key: K, value: V, metadata: Metadata): D = {
+      val innerResult = inner.filterAndConvert(key, value, metadata)
+      if (innerResult == null) null.asInstanceOf[D]
+      else {
+         outer.filterAndConvert(key, innerResult, metadata)
+      }
+   }
+}
+
+object WrappedFilter {
+   def apply[K, V, C, D](outer: InjectableFilter[K, C, D],
+                         inner: InjectableFilter[K, V, C]): WrappedFilter[K, V, C, D] =
+      new WrappedFilter(outer, inner)
+}
+
+class WrappedFilterExternalizer[K, V, C, D] extends AbstractExternalizer[WrappedFilter[K, V, C, D]] {
+   override def getTypeClasses = setAsJavaSet(Set(classOf[WrappedFilter[K, V, C, D]]))
+
+   override def writeObject(output: ObjectOutput, obj: WrappedFilter[K, V, C, D]) = {
+      output.writeObject(obj.outer)
+      output.writeObject(obj.inner)
+   }
+
+   override def readObject(input: ObjectInput): WrappedFilter[K, V, C, D] = {
+      val outer = input.readObject().asInstanceOf[InjectableFilter[K, C, D]]
+      val inner = input.readObject().asInstanceOf[InjectableFilter[K, V, C]]
+      WrappedFilter(outer, inner)
+   }
+}
+
+class UnmarshallKeyValueFilterConverter[K, V, C](val delegate: KeyValueFilterConverter[K, V, C], val marshallerClass: Class[_ <: Marshaller]) extends InjectableFilter[Bytes, Bytes, Bytes] {
```

Comment:
I'm afraid we need both. In the events case the unmarshall wraps a ```CacheEventFilterConverter``` while here it's on top of a ```KeyValueFilterConverter```